src/posix/platform/netif.cpp

```diff
@@ -115,31 +240,185 @@ enum
+    size_t         i;
+    ascii_buffer[0] = 0;
+    for (i = 0; i < len; i++)
+    {
+        uint16_t val = (*buf) & 0x00FF;
+        buf++;
+
+        ascii_buffer[i % 8]     = isprint(val) ? (char)val : '.';
+        ascii_buffer[i % 8 + 1] = 0;
+
+        if ((i % 8) == 0)
+        {
+            if (i != 0)
+            {
+                fprintf(stderr, "\r\n");
+            }
+            fprintf(stderr, "\t");
+        }
+        if ((i % 4) == 0)
+            fprintf(stderr, "\t");
+
+        fprintf(stderr, "%02hX ", val);
+
+        if (((i + 1) % 8) == 0)
+        {
+            fprintf(stderr, "\t%s", ascii_buffer);
+            ascii_buffer[0] = 0;
+        }
+    }
+    while ((i % 8) != 0)
+    {
+        if ((i % 4) == 0)
+        {
+            fprintf(stderr, "\t");
+        }
+
+        fprintf(stderr, "   ");
+        i++;
+    }
+    fprintf(stderr, "\t%s\r\n", ascii_buffer);
+}
+
+#if !defined(__linux__)
+static bool unicastAddressIsSubscribed(otInstance *aInstance, const otNetifAddress *netAddr)
+{
+    const otNetifAddress *address = otIp6GetUnicastAddresses(aInstance);
+
+    while (address != NULL)
+    {
+        if (memcmp(address->mAddress.mFields.m8, netAddr->mAddress.mFields.m8, sizeof(address->mAddress.mFields.m8)) ==
+            0)
+        {
+            return true;
+        }
+
+        address = address->mNext;
+    }
+
+    return false;
+}
+#endif
+
+#if defined(__APPLE__) || defined(__NetBSD__) || defined(__FreeBSD__)
+static inline void apply_mask(struct in6_addr *address, uint8_t mask)
+{
+    if (mask > 128)
+    {
+        mask = 128;
+    }
+
+    memset((void *)(address->s6_addr + ((mask + 7) / 8)), 0, 16 - ((mask + 7) / 8));
+
+    if (mask % 8)
+    {
+        address->s6_addr[mask / 8] &= ~(0xFF >> (mask % 8));
+    }
+}
+
+static unsigned int calculate_prefix_len(const struct sockaddr_in6 *sa6)
+{
+    unsigned int result = 0;
+
+    unsigned int i;
+
+    for (i = 0; i < sizeof(sa6->sin6_addr.s6_addr); i++)
+    {
+        if (sa6->sin6_addr.s6_addr[i] == 0xFF)
+        {
+            result += 8;
+            continue;
+        }
+        else
+        {
+            int     j;
+            uint8_t byte = sa6->sin6_addr.s6_addr[i];
+
+            for (j = 7; j >= 0; j--)
+            {
+                if (byte & (1 << j))
+                    result++;
+                else
+                    break;
+            }
+
+            break;
+        }
+    }
+
+    return result;
+}
+#endif
+
 static void UpdateUnicast(otInstance *aInstance, const otIp6Address &aAddress, uint8_t aPrefixLength, bool aIsAdded)
 {
-    struct in6_ifreq ifr6;
-    otError          error = OT_ERROR_NONE;
+    otError error = OT_ERROR_NONE;
 
     assert(sInstance == aInstance);
 
-    VerifyOrExit(sIpFd > 0, error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(sIpFd >= 0, error = OT_ERROR_INVALID_STATE);
 
-    memcpy(&ifr6.ifr6_addr, &aAddress, sizeof(ifr6.ifr6_addr));
+#if defined(__linux__)
```

Comment:
I thought about that, but I'm really nervous about that.  There are some functions that are indeed different, but for the most part, big chunks of the code are very similar.  I'd like to avoid significant duplication because duplicated code always diverges, which means means that at least SOME of the code needs to be in a file that's common.  I worry that if the mac/BSD implementation is separate, people will carry one of the implementations forward without the other.  By putting it here, it's easier to see when doing something Linux-specific (or Mac-specific) will break something.